syntax = "proto3";
package model.protos;
import "google/protobuf/descriptor.proto";

enum Qos {
  UNSET = 0;
  AT_MOST_ONCE = 1;
  AT_LEAST_ONCE = 2;
  EXACTLY_ONCE = 3;
}

enum Operator {
  SUM = 0;
  AVG = 1;
}

message HomieNode {
  string id = 1;
  string name = 2;
  string type = 3;
  bool no_retain = 4;
  Qos qos = 5;
}

message DerivedField {
  Operator operator = 1;
  string field_name = 2;
  string display_name = 3;
  string node = 4;
  repeated string fields = 6;
  optional string unit = 7;
}

extend google.protobuf.MessageOptions {
  repeated HomieNode homie_node = 50001;
  repeated DerivedField derived_field = 50002;
}

message MappingOptions {
  // divide the incoming value by this
  optional uint32 divisor = 1;
  // unit of the data
  optional string unit = 2;
  // special converter to show the value in a nicer way (e.g. convert minutes into HH:MM)
  optional string converter = 3;

  // mapping target name in homie (will be auto-generated by the field name, but can be overridden by this value)
  optional string id = 4;
  // the homie node this field should be added to as property
  optional string node = 5;
  // optional name for this property
  optional string display_name = 6;
}

extend google.protobuf.FieldOptions {
    optional MappingOptions mapping_options = 50002;
  }
  